### ISSUE TITLE
Optimize-DbContext: same problem as many times before: Exceeds maximum filename length

### ISSUE THREAD

**Issue by henrikdahl8240** [dotnet/efcore#26896]
Many times earlier, Microsoft's development tools have had the specific weakness that some limit has been exceeded, also in case of EF Core.

It can be an idea to create a deep hierarchy using CASE which exceeds the maximum height, 10, which SQL Server supports.

Maximum length of filename for automatically generated filenames, which Windows Workflow Fundation did.

Now Optimize-DbContext does exactly the same, creates some enormously long filenames, so the command fails at some time because the generated filename exceeds the limit.

Is there any idea in creating these countless tiny files and not just one as it's not possible to update the created compiled model but only to create a new one from scratch.

I think you should either:
*) Create only one file with all the stuff.
*) Create files having names like f1, f2, f3 etc., then the lenght shoould not be exceeded.

The current approach with countless tiny files is generally a very inefficient way to use the filesystem.

Couldn't you instruct your staff, that as they dynamically create something in some domain having specific limits, then the staff should ensure, that these limits will not be exceeded. It can not be so difficult to adhere to.

Now, due to your decision, this is again something which is only usable in tiny demo-cases. A problem which has characterized EF for all the years as you can read about in countless posts, particularly in scope of EF before CORE.

I am sorry for my rather negative tone, but I consider the persistent approach to denote complete carelessness. Can it really be correct it must be a problem again and again?

So, due to your decision, I cannot use this at all. Probably it's the same concerning many other people.

**Comment by ajcvickers** [comment]
@henrikdahl8240 Can you post an example of the problematic filenames?

**Comment by henrikdahl8240** [comment]
@ajcvickers Yes, of course. I have replaced two business concepts with Xxx... and Y... where X... and Y... obviously have the same length as the original business concepts.
The filename is:
BusinessPropertyInstance__Literal__Timeline_Generation2XxxxxxxXxxxxxx_YyyyyyyyyyyyyyyyyyyyyInventoryBusinessPropertyInstance__Literal__Timeline_Entry_Generation2XxxxxxxXxxxxxx_YyyyyyyyyyyyyyyyyyyyyInventoryDateTimeOffset_SimpleDateTimeOffset_SimpleEntityType.cs

The length of the filename is 261 characters.

The class in question has four generic types. The names you generate include the actual types, I mean like if you have:
class C<T1, T2, T3, T4> and the mapping is done for C<A, B, C, D>, then the filename becomes CABCD.cs.

It is obviously how it should be in order to make the filename unique. It is just a decision which I would say you cannot draw as there is such a limitation on the maximum length of filename because it will cause, that arbitrary limitations are being imposed on the names you can use for classes in your model. I suppose you agree, it's not and opinion which EF Core should have. EF Core should obviously not have any opinion on names you may use for your classes in your model.

**Comment by henrikdahl8240** [comment]
@ajcvickers  So as far as I understand, you have decomposed all this EF Core stuff into separate components provided by dependency injection at the given run.

Does it mean, that I can just provide my own provider for producing the filenames and just make it to provide file1.cs, file2.cs etc.instead of the names you have chosen by your default provider?

**Comment by ajcvickers** [comment]
@henrikdahl8240 Possibly. Depends on the specific service boundaries for this case.
